src/main/scala/com/microsoft/ml/spark/vw/VowpalWabbitFeaturizer.scala

```diff
@@ -0,0 +1,141 @@
+
+        m.valueType match {
+          case StringType => new MapStringFeaturizer(idx, name, namespaceHash, getMask)
+          case DoubleType => new MapFeaturizer[Double](idx, name, namespaceHash, getMask, v => v)
+          case FloatType => new MapFeaturizer[Float](idx, name, namespaceHash, getMask, v => v.toDouble)
+          case IntegerType => new MapFeaturizer[Int](idx, name, namespaceHash, getMask, v => v.toDouble)
+          case LongType => new MapFeaturizer[Long](idx, name, namespaceHash, getMask, v => v.toDouble)
+          case ShortType => new MapFeaturizer[Short](idx, name, namespaceHash, getMask, v => v.toDouble)
+          case ByteType => new MapFeaturizer[Byte](idx, name, namespaceHash, getMask, v => v.toDouble)
+          case _ => throw new RuntimeException(s"Unsupported map value type: ${dataType}")
+        }
+      }
+      case m: Any =>
+        if (m == VectorType) // unfortunately the type is private
+          new VectorFeaturizer(idx, getMask)
+        else
+          throw new RuntimeException(s"Unsupported data type: ${dataType}")
+    }
+  }
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val inputColsList = getAllInputCols
+    val namespaceHash: Int = VowpalWabbitMurmur.hash(this.getOutputCol, this.getSeed)
+
+    val fieldSubset = dataset.schema.fields
+      .filter(f => inputColsList.contains(f.name))
+
+    val featurizers: Array[Featurizer] = fieldSubset.zipWithIndex
+      .map { case (field, idx) => getFeaturizer(field.name, field.dataType, idx, namespaceHash) }
+
+        // TODO: list types
+        // BinaryType
+        // CalendarIntervalType
+        // DateType
+        // NullType
+        // TimestampType
+        // getStruct
+
+    val mode = udf((r: Row) => {
+      val indices = ArrayBuilder.make[Int]
+      val values = ArrayBuilder.make[Double]
+
+      // educated guess on size
+      indices.sizeHint(featurizers.length)
+      values.sizeHint(featurizers.length)
+
+        // apply all featurizers
+      for (f <- featurizers)
+        if (!r.isNullAt(f.fieldIdx))
+          f.featurize(r, indices, values)
+
+      // sort by indices and remove duplicate values
+      // Warning:
+      //   - due to SparseVector limitations (which doesn't allow duplicates) we need filter
+      //   - VW command line allows for duplicate features with different values (just updates twice)
+      val (indicesSorted, valuesSorted) = VectorUtils.sortAndDistinct(indices.result, values.result, getSumCollisions)
+
+      Vectors.sparse(1 << getNumBits, indicesSorted, valuesSorted)
+    })
+
+    dataset.toDF.withColumn(getOutputCol, mode.apply(struct(fieldSubset.map(f => col(f.name)): _*)))
+  }
+
+  override def copy(extra: ParamMap): VowpalWabbitFeaturizer = defaultCopy(extra)
+
+  override def transformSchema(schema: StructType): StructType = {
+    val fieldNames = schema.fields.map(_.name)
+    for (f <- getAllInputCols)
```

Comment:
getAllInputCols.foreach( { f => if (!fieldNames.contains(f)) throw new IllegalArgumentException("missing input column " + f) } )

Reply:
maybe it's just me, but I don't feel this improves readability. if you have a strong opinion on this as it's not to scala style enough